proposed/swift-interop.md

```diff
@@ -152,33 +152,47 @@ If possible, Swift tuples should be represented as `ValueTuple`s in .NET. If thi
+Importing Swift into .NET is done through the following steps:
+
+1. The tool analyzes a source library `.swift` and maps entry points to mangled names.
+2. Then, the tool aggregates the public API; it take in a `.swiftmodule` file (similar to a C header file) generated by the Swift compiler.
+3. For cases where direct P/Invoke from C# into Swift is not possible, the tool generates source code `.swift` wrappers and compiles them.
+4. Subsequently, the tool aggregates the public API for the generated wrappers.
+5. Finally, the tool generates C# bindings source code based on the aggregated public API and compiles them.
 
-#####  Exporting .NET to Swift
+##### Exporting .NET to Swift
 
 There are two components to exporting .NET to Swift: Implementing existing Swift types in .NET and passing instances of those types to Swift, and exposing novel types from .NET code to Swift code to be created from Swift. Exposing novel types from .NET code to Swift code is considered out of scope at this time.
 
 For implementing existing Swift types in .NET, we will require one of the following tooling options:
 
 1. A Roslyn source generator to generate any supporting code needed to produce any required metadata, such as type metadata and witness tables, to pass instances of Swift-type-implementing .NET types defined in the current project to Swift.
-2. An IL-post-processing tool to generate the supporting code and metadata from the compiled assembly.
+2. An IL-post-processing tool to generate the supporting code and metadata from the compiled assembly. 
+
+If we were to use an IL-post-processing tool here, we would break Hot Reload in assemblies that implement Swift types, even for .NET-only code, due to introducing new tokens that the Hot Reload "client" (aka Roslyn) does not know about. As a result, we should prefer the Roslyn source generator approach. The existing tool [supports exporting .NET types to Swift](https://github.com/xamarin/binding-tools-for-swift/blob/main/docs/FutureIdeas.md#exposing-c-to-swift) that can work as a source generator.
 
-If we were to use an IL-post-processing tool here, we would break Hot Reload in assemblies that implement Swift types, even for .NET-only code, due to introducing new tokens that the Hot Reload "client" (aka Roslyn) does not know about. As a result, we should prefer the Roslyn source generator approach.
+For a C# class that can be inherited gets implemented in C#, the tool generates a mirror class in Swift and overrides it with another class that vectors all the calls through a simulated vtable. This class implements the constructor and overrides the virtual methods by sending calls to a function pointer which ends up in C#.
+
+##### Limitations of the tool
+
+The current version of the projection tooling has limitations and, in some cases, requires Swift wrappers in addition to C# bindings. Here are some examples:
```

Comment:
Right, the current version of the tool as a prototype that took number of shortcuts and that has number of missing parts. It is not necessary to list them in this design doc. This design doc should be focused on the shape we want to ship as the first version. I think all these limitations need to be addressed for the first shipping version. I think it would be a good idea to fix the redundant input problem too.

Reply:
> I think all these limitations need to be addressed for the first shipping version. I think it would be a good idea to fix the redundant input problem too.

I believe it's essential to provide a detailed mapping of types along with their constraints and limitations, which could fall under `### Projecting Swift into .NET`. This will help us in defining the scope of functionalities the interop.

> Have you done analysis of how much of the prototype code is in the shape that we really like vs. how much of it will need to be redone?

This will definitely help in our decision-making process, I will make such analysis. My understanding is that the Swift wrappers represent a minority of the code compared to the code that is in the shape that we really want. 

I'm aware of the sunk cost fallacy, but I'd like to approach this pragmatically. On one hand, the existing tool has a lot of features already implemented which are in a good shape and has robust test coverage. On the other hand, starting from scratch and selectively incorporating parts of the prototype that we like may allow us to move faster and allow for easier decision-making.

Reply:
With regards to the state of the current projection code - I don't think that starting from ground-zero is a good idea.
The code is fairly modular in that there are sections that don't have to change and have already solved hard problems - such as parsing swiftinterface files, detangling swift entry points, representing swift types, authoring C# and swift code, maintaining a database of already bound types and types in the module being bound. One of the consistent problems with swift is that there are so many special cases in the language. The ugliest portions of BtFS have to do with handling all these special cases. For example, calling conventions on struct methods use one convention unless the type is marked `@frozen` then it uses a different convention. Passing structs into methods depends on both the size of the struct as well as whether it's frozen. Closures get broken up into special cases for each of the following categories: escaping/non-escaping, throws or no throw, returns a value or doesn't return a value, async or not, which is 16 unique cases, all of which get handled slightly differently.
The point is that there is a lot of wisdom already tied up in the tool, however, during test runs, I've spent time thinking about how to make the organization and handling of those special cases better and it can be done in a way that replaces the ugly portions of the code entirely. So rather than replacing what's there, the thought should instead be around how we can refactor it instead.

Reply:
After offline sync, we agreed that the components of the tool should be moved under the dotnet organization for projection tooling. The idea is to transfer the parts of the prototype that we like from https://github.com/xamarin/binding-tools-for-swift to this new repository/directory and through that process perform a design and code review to improve it. Here is the proposal: https://github.com/dotnet/runtime/pull/97208

We want to avoid a big burden in terms of compliance and integration with a new repository. Also, in this way it is closer to the runtime and the integration/shipping will be easier.

Reply:
If we're going to move the tools into the dotnet/runtime repository, I feel very strongly that we should use the core of the interop source-gen tooling to implement some of the logic for marshalling. We found some success with it in the WASM JS interop space and it enables us to avoid re-solving some problems again and again.

Reply:
You already mentioned this in a comment below.